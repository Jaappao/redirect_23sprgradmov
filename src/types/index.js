// @flow
import moment from 'moment'

export type Time = {|
	+h: number,
	+m: number,
|}

export type PeriodInfo = {|
	+period: number,
	+start: Time,
	+end: Time,
|}

export type PeriodStatusBefore = {
	type: 'before',
	fromNowStr: string,
}

export type PeriodStatusProgress = {
	type: 'progress',
	progress: number,
	rate: number,
}

export type PeriodStatusFinish = {
	type: 'finish',
}
export type PeriodStatus =
	| PeriodStatusBefore
	| PeriodStatusProgress
	| PeriodStatusFinish
	| null

export type Period = {|
	+info: PeriodInfo,
	status: PeriodStatus,
	+start: moment,
	+end: moment,
|}
