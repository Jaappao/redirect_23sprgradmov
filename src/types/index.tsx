export type Time = {
	h: number
	m: number
}

export type PeriodInfo = {
	period: string
	start: Time
	end: Time
}

export type PeriodStatusType = 'before' | 'progress' | 'finish'
export type PeriodStatusBefore = {
	type: 'before'
}

export type PeriodStatusProgress = {
	type: 'progress'
	progress: number
	rate: number
}

export type PeriodStatusFinish = {
	type: 'finish'
}
export type PeriodStatus =
	| PeriodStatusBefore
	| PeriodStatusProgress
	| PeriodStatusFinish
	| null

export type Period = {
	readonly info: PeriodInfo
	status: PeriodStatus
}

export type Profile = {
	periods: {
		[key: string]: PeriodInfo
	}
}

export type TimeResponse = {
	base: Profile
	d: Profile
	j: Profile
}
