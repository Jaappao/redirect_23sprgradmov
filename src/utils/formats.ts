import { Book, Study, Time } from '../types'

export const pad2 = (n: number) => `${n}`.padStart(2, '0')

export const pad = (t: Time) => {
	return `${pad2(t.h)}:${pad2(t.m)}`
}

export const hex2Bools = (str: string): boolean[] => {
	return str
		.split('')
		.reduce(
			(p: boolean[], v) =>
				p.concat(
					parseInt(v, 16)
						.toString(2)
						.padStart(4, '0')
						.split('')
						.map(Number)
						.map(Boolean),
				),
			[],
		)
}
export const bools2Hex = (bools: boolean[]) => {
	return arrayChunk(bools, 4)
		.map((bls) => {
			return parseInt(bls.map(Number).map(String).join(''), 2).toString(16)
		})
		.join('')
}

function arrayChunk<T>(arr: T[], chunk: number): T[][] {
	const res: T[][] = []

	arr.forEach((v, i) => {
		const k = Math.floor(i / chunk)

		if (!res[k]) res[k] = []
		res[k].push(v)
	})
	return res
}
const wds = [...Array(7).keys()]

export function encodeStudy(study: Study, periodKeys: string[]): string {
	const bools = wds
		.map((wi) => study[wi] || {})
		.reduce((p: boolean[], c) => p.concat(periodKeys.map((pk) => !!c[pk])), [])

	return bools2Hex(bools)
}

export function decodeStudy(key: string, periodKeys: string[]): Study {
	const bools = hex2Bools(key)
	const study: Study = {}

	wds.forEach((w, wi) => {
		study[w] = {}
		periodKeys.forEach((pk, pi) => {
			study[w][pk] = bools[wi * periodKeys.length + pi] || false
		})
	})
	return study
}

export const bookId = (book: Book) => bookIdFormat(book.pid, book.label)
export const bookIdFormat = (pid: string, bid: string) => `${pid}/${bid}`
