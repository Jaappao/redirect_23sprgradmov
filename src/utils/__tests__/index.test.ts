import 'jest-styled-components'
import { encodeStudy, decodeStudy, hex2Bools } from '../formats'
import { Study } from '../../types'

const study: Study = {
	'2': { '1': true, '2': true },
	'4': { '3': false },
	'5': { '1': true, '2': true, '4': false, '5': false },
	'6': { '1': false, '2': false, '3': false, D6: true },
}
const key = '0000c00000c001'
const pks = ['1', '2', '3', '4', '5', '6', 'D5', 'D6']

test('encodeStudy', () => {
	expect(encodeStudy(study, pks)).toStrictEqual(key)
})

test('decodeStudy', () => {
	expect(decodeStudy(key, pks)).toMatchSnapshot()
})

test('hex2Bools', () => {
	expect(hex2Bools('00af')).toStrictEqual([
		false,
		false,
		false,
		false,
		false,
		false,
		false,
		false,
		true,
		false,
		true,
		false,
		true,
		true,
		true,
		true,
	])
})
