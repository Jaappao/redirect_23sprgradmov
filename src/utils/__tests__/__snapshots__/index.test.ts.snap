// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decodeStudy 1`] = `
Object {
  "0": Object {
    "1": false,
    "2": false,
    "3": false,
    "4": false,
    "5": false,
    "6": false,
    "D5": false,
    "D6": false,
  },
  "1": Object {
    "1": false,
    "2": false,
    "3": false,
    "4": false,
    "5": false,
    "6": false,
    "D5": false,
    "D6": false,
  },
  "2": Object {
    "1": true,
    "2": true,
    "3": false,
    "4": false,
    "5": false,
    "6": false,
    "D5": false,
    "D6": false,
  },
  "3": Object {
    "1": false,
    "2": false,
    "3": false,
    "4": false,
    "5": false,
    "6": false,
    "D5": false,
    "D6": false,
  },
  "4": Object {
    "1": false,
    "2": false,
    "3": false,
    "4": false,
    "5": false,
    "6": false,
    "D5": false,
    "D6": false,
  },
  "5": Object {
    "1": true,
    "2": true,
    "3": false,
    "4": false,
    "5": false,
    "6": false,
    "D5": false,
    "D6": false,
  },
  "6": Object {
    "1": false,
    "2": false,
    "3": false,
    "4": false,
    "5": false,
    "6": false,
    "D5": false,
    "D6": true,
  },
}
`;
